### ISSUE TITLE
FP32 gives good results but int8 predicts only single class

### ISSUE THREAD

**Issue by vemqos99** [tensorflow/tensorflow#89438]
### 1. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 22.04
- TensorFlow installation (pip package or built from source): pip package
- TensorFlow library (version, if pip package or github SHA, if built from source): 2.15.0

### 2. Code

Provide code to help us reproduce your issues using one of the following options: Standard code to convert from fp32 int8 taken directly from website. The representative dataset is the validation set

The model used is Nanodet  for 2 classes. used with images of size 160x160. Each class has atleat 2000 images. Total training set size is 5500. validation set size is 1800.
nanodet model converted to onnx then converted to tflite.
All fp32 models give the same results as below. int8 model results is skewed.

[nanodet_detection_results_fp32.txt](https://github.com/user-attachments/files/19312938/nanodet_detection_results_fp32.txt)
[nanodet_detection_results_int8.txt](https://github.com/user-attachments/files/19312939/nanodet_detection_results_int8.txt)

As you can see from the results, int8 model predicts only single class.
In the detection results, 1-30 is class0, 31-60 is class1, rest should not detect and class with high confidence.

This is the .py code for int8. for fp32 the items related to quantization is removed
import cv2
import numpy as np
import tensorflow as tf
import math

# Thresholds
CONF_THRESH = 0.5  # Confidence threshold
IOU_THRESH = 0.1   # IoU threshold for NMS
REG_MAX = 7

# Define colors for visualization
_COLORS = np.array([
    [0.000, 0.447, 0.741],  # Class 0 (blue)
    [0.850, 0.325, 0.098],  # Class 1 (orange)
])

# Load INT8 TFLite model
def load_tflite_model(model_path):
    interpreter = tf.lite.Interpreter(model_path=model_path)
    interpreter.allocate_tensors()
    return interpreter

# Preprocess image for INT8 model
def preprocess_image(image_path, input_details):
    img = cv2.imread(image_path, cv2.IMREAD_COLOR)
    orig_shape = img.shape[:2]
    img = cv2.resize(img, (160, 160))

    # Normalize to INT8 scale
    input_dtype = input_details[0]['dtype']
    input_scale, input_zero_point = input_details[0]['quantization']
    
    img = img.astype(np.float32) / 255.0  # Normalize to [0,1]
    img = (img / input_scale + input_zero_point).astype(input_dtype)  # Quantize
    img = np.expand_dims(img, axis=0)  # NHWC format
    return img, orig_shape

# Generate grid center priors
def generate_grid_center_priors(input_height, input_width, strides):
    center_priors = []
    for stride in strides:
        feat_w = math.ceil(input_width / stride)
        feat_h = math.ceil(input_height / stride)
        for y in range(feat_h):
            for x in range(feat_w):
                center_priors.append([x, y, stride])
    return center_priors

# Softmax activation function
def activation_function_softmax(src):
    alpha = max(src)
    denominator = sum(np.exp(src - alpha))
    return np.exp(src - alpha) / denominator

# Decode distributional predictions to bounding boxes
def disPred2Bbox(dfl_det, label, score, x, y, stride):
    ct_x = x * stride
    ct_y = y * stride
    dis_pred = []
    for i in range(4):
        dis = sum(j * activation_function_softmax(dfl_det[i * (REG_MAX + 1):(i + 1) * (REG_MAX + 1)])[j] for j in range(REG_MAX + 1))
        dis *= stride
        dis_pred.append(dis)
    xmin, ymin = max(ct_x - dis_pred[0], 0), max(ct_y - dis_pred[1], 0)
    xmax, ymax = min(ct_x + dis_pred[2], 160), min(ct_y + dis_pred[3], 160)
    return [xmin, ymin, xmax, ymax, score, label]

# Decode NanoDet output (with INT8 dequantization)
def decode_infer(pred, center_priors, score_threshold, output_details, num_classes=2):
    output_scale, output_zero_point = output_details[0]['quantization']
    pred = (pred.astype(np.float32) - output_zero_point) * output_scale  # Dequantize
    pred = pred.flatten()
    results = {i: [] for i in range(num_classes)}
    num_channels = num_classes + (REG_MAX + 1) * 4
    for idx, center_prior in enumerate(center_priors):
        ct_x, ct_y, stride = center_prior
        scores = pred[idx * num_channels:idx * num_channels + num_classes]
        cur_label = np.argmax(scores)
        score = scores[cur_label]
        if score > score_threshold:
            bbox_pred = pred[idx * num_channels + num_classes:]
            results[cur_label].append(disPred2Bbox(bbox_pred, cur_label, score, ct_x, ct_y, stride))
    return results

# Non-Maximum Suppression
def nms(dets, thresh):
    if len(dets) == 0:
        return []
    x1, y1, x2, y2, scores = dets[:, 0], dets[:, 1], dets[:, 2], dets[:, 3], dets[:, 4]
    areas = (x2 - x1 + 1) * (y2 - y1 + 1)
    order = scores.argsort()[::-1]
    keep = []
    while order.size > 0:
        i = order[0]
        keep.append(i)
        xx1 = np.maximum(x1[i], x1[order[1:]])
        yy1 = np.maximum(y1[i], y1[order[1:]])
        xx2 = np.minimum(x2[i], x2[order[1:]])
        yy2 = np.minimum(y2[i], y2[order[1:]])
        w = np.maximum(0.0, xx2 - xx1 + 1)
        h = np.maximum(0.0, yy2 - yy1 + 1)
        inter = w * h
        ovr = inter / (areas[i] + areas[order[1:]] - inter)
        order = order[np.where(ovr <= thresh)[0] + 1]
    return keep

# Run inference and visualize results
def run_inference(model_path, image_path):
    interpreter = load_tflite_model(model_path)
    input_details, output_details = interpreter.get_input_details(), interpreter.get_output_details()
    
    image, orig_shape = preprocess_image(image_path, input_details)
    
    interpreter.set_tensor(input_details[0]['index'], image)
    interpreter.invoke()
    
    dets = interpreter.get_tensor(output_details[0]['index'])
    center_priors = generate_grid_center_priors(160, 160, [8, 16, 32])
    preds = decode_infer(dets, center_priors, CONF_THRESH, output_details)
    
    results = []
    for key, pred in preds.items():
        idxs = nms(np.array(pred), IOU_THRESH)
        results.extend(pred[idx] for idx in idxs)
    
    # Map bbox to original image size
    src_w, src_h = orig_shape[1], orig_shape[0]
    img = cv2.imread(image_path)
    for result in results:
        x1, y1, x2, y2, score, label = result
        x1, y1, x2, y2 = int(x1 * src_w / 160), int(y1 * src_h / 160), int(x2 * src_w / 160), int(y2 * src_h / 160)
        color = (_COLORS[label] * 255).astype(np.uint8).tolist()
        cv2.rectangle(img, (x1, y1), (x2, y2), color, 2)
        cv2.putText(img, f"Class {label}: {score:.2f}", (x1, y1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, color, 2)
    
    cv2.imwrite("output.jpg", img)
    return results

if __name__ == "__main__":
    model_file = "nanodet_int8.tflite"
    image_file = "D:\\regressionTest\\data\\carLP\\IMG_008.jpg"
    results = run_inference(model_file, image_file)
    print(results)
